### DISCUSSION TITLE
GINE does not work beyond the first convolution

### DISCUSSION THREAD

**Discussion by mdanb** [pyg-team/pytorch_geometric#2171] [Q&A] [answered]
I am trying to use GINE conv for a weighted graph where the node features are 1 dimensional, and the `edge_attr` (i.e edge weights) of my `Data` object are also 1D. My hidden dimensions are 16D. The following shows my code for my `forward` method, which uses GINE convs. 
   ```
   def forward(self, data):
        x, edge_index, edge_attr = data.x, data.edge_index, data.edge_attr
        for i in range(self.num_layers):
            x = self.convs[i](x, edge_index, edge_attr) # convs is just a list of GINE convs, where the first GINE conv is 1x16 and the rest are 16 x 16
            x = F.relu(x)
            x = F.dropout(x, p=self.dropout, training=self.training)
            if not i == self.num_layers - 1:
                x = self.lns[i](x) # lns is just a list of LayerNorms with hidden dim = 16

        x = global_add_pool(x)

        x = self.post_mp(x) # this is just an MLP

        return F.log_softmax(x, dim=1)
  ```

The first iteration through the loop works, but beyond the first I get the error:
```
  File "/home/mdanb/.local/lib/python3.8/site-packages/torch_geometric/nn/conv/gin_conv.py", line 109, in forward
    assert x.size(-1) == edge_attr.size(-1)
AssertionError
```
`x.size(-1)` is 16 and `edge_attr.size(-1)` is 1
The error makes sense I guess since the update/aggregation for GINE involves adding up the node embeddings and edge embeddings (as opposed to GCN where we consider weights as scalars by which we scale node embeddings). So how exactly should I incorporate processing the edge weights in the code above?

**Comment by rusty1s** [comment, marked as answer]
In general, you need to take care about this yourself. I suggest to use a `Linear` layer to transform edge features to the appropriate size, *e.g.*, `Linear(num_edge_features, num_hidden_channels)`:
```python
x = self.convs[i](x, edge_index, self.edge_lins[i](edge_attr))
```

**Reply by mdanb** [reply]
I see, thank you! I'm assuming you meant doing this:
```
x = self.convs[i](x, edge_index, edge_attr)
edge_attr = self.edge_lins[i](edge_attr)
```
(The other one didn't work because the edge attributes were being transformed before the convolution so the node attributes were still in the "old" dimension)
